internal/pkg/manifest/workload_applyenv_test.go

```diff
@@ -0,0 +1,1846 @@
+package manifest
+
+import (
+	"testing"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/stretchr/testify/require"
+)
+
+func Test_ApplyEnv_Image(t *testing.T) {
+	testCases := map[string]struct {
+		inSvc  func(svc *LoadBalancedWebService)
+		wanted func(svc *LoadBalancedWebService)
+	}{
+		"exclusive fields: build overridden if location is not nil": {
+			inSvc: func(svc *LoadBalancedWebService) {
+				svc.ImageConfig.Build = BuildArgsOrString{
+					BuildString: aws.String("mockBuild"),
+				}
+				svc.Environments["test"].ImageConfig.Location = aws.String("mockLocation")
+			},
+			wanted: func(svc *LoadBalancedWebService) {
+				svc.ImageConfig.Location = aws.String("mockLocation")
+				svc.ImageConfig.Build = BuildArgsOrString{
+					BuildString: nil,
+					BuildArgs:   DockerBuildArgs{},
+				}
+			},
+		},
+		"exclusive fields: location overridden if build is not nil": {
+			inSvc: func(svc *LoadBalancedWebService) {
+				svc.ImageConfig.Location = aws.String("mockLocation")
+				svc.Environments["test"].ImageConfig.Build = BuildArgsOrString{
+					BuildArgs: DockerBuildArgs{
+						Dockerfile: aws.String("mockDockerfile"),
+					},
+				}
+			},
+			wanted: func(svc *LoadBalancedWebService) {
+				svc.ImageConfig.Location = nil
+				svc.ImageConfig.Build = BuildArgsOrString{
+					BuildString: nil,
+					BuildArgs: DockerBuildArgs{
+						Dockerfile: aws.String("mockDockerfile"),
+					},
+				}
+			},
+		},
+		"build overridden": {
+			inSvc: func(svc *LoadBalancedWebService) {
+				svc.ImageConfig.Build = BuildArgsOrString{
+					BuildString: aws.String("mockBuild"),
+				}
+				svc.Environments["test"].ImageConfig.Build = BuildArgsOrString{
+					BuildString: aws.String("mockBuildTest"),
+				}
+			},
+			wanted: func(svc *LoadBalancedWebService) {
+				svc.ImageConfig.Build = BuildArgsOrString{
+					BuildString: aws.String("mockBuildTest"),
+				}
+			},
+		},
+		"build not overridden": {
+			inSvc: func(svc *LoadBalancedWebService) {
+				svc.ImageConfig.Build = BuildArgsOrString{
+					BuildString: aws.String("mockBuild"),
+				}
+				svc.Environments["test"].ImageConfig.Build = BuildArgsOrString{}
```

Comment:
nit: can we remove this

Reply:
I think it makes sense to update it to
```suggestion
				svc.Environments["test"].ImageConfig= ImageConfig{}
```
This is because sometimes one of the parent field is `nil` and it makes a difference to the test result how far above the struct is initialized :(

Reply:
To better explain this: https://github.com/aws/copilot-cli/blob/ad12a5783597574036da61348af8d515cc2b12a1/internal/pkg/manifest/workload_applyenv_test.go#L1810

When the user specify
```yaml
network:
   vpc:
       placement: some_placement
```
The VPC struct will unmarshaled as 
```go
{
   placement: aws.String("some_placement")
   security_groups: nil
}
```
which results in the failed case in https://github.com/aws/copilot-cli/blob/ad12a5783597574036da61348af8d515cc2b12a1/internal/pkg/manifest/workload_applyenv_test.go#L1810.

However, if the network field is not specified at all, it will be `nil`, and `security_groups` won't be overridden because `Network` is a pointer type and unmarshaled as `nil`.

Reply:
I think I understand it! So if the parent field is a struct (like "vpc") we want to make sure that `security_groups` is maintained in the following scenario:
```yaml
network:
  vpc:
    security_groups: ['asdf']

environments:
  test:
     network:
        vpc:
          placement: 'private'
```
that's why you test with empty structs and it makes a lot of sense.


My recommendation is two things:
1. Rename the file to `applyenv_test.go`.
2. Add a big block comment at the top of the file explaining how to write tests here. The guideline should essentially tell me: "OK I've added a new type to the manifest, how should I write test cases for it based on the type"